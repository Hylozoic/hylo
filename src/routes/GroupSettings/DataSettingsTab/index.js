import component from './DataSettingsTab'

export default component
