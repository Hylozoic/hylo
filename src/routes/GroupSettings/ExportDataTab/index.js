import ExportDataTab from './ExportDataTab'
import connector from './ExportDataTab.connector'
export default connector(ExportDataTab)
