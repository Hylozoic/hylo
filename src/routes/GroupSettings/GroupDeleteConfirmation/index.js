import GroupDeleteConfirmation from './GroupDeleteConfirmation'
import connector from './GroupDeleteConfirmation.connector'
export default connector(GroupDeleteConfirmation)
