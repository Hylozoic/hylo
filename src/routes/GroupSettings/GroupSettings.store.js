export const MODULE_NAME = 'GroupSettings'
export const REGENERATE_ACCESS_CODE = `${MODULE_NAME}/REGENERATE_ACCESS_CODE`
export const FETCH_GROUP_SETTINGS = `${MODULE_NAME}/FETCH_GROUP_SETTINGS`
export const UPDATE_GROUP_SETTINGS = `${MODULE_NAME}/UPDATE_GROUP_SETTINGS`
export const UPDATE_GROUP_SETTINGS_PENDING = UPDATE_GROUP_SETTINGS + '_PENDING'
export const DELETE_GROUP = `${MODULE_NAME}/DELETE_GROUP`

export function orderFromSort (sortBy) {
  if (sortBy === 'name') return 'asc'
  return 'desc'
}

export function fetchGroupSettings (slug) {
  return {
    type: FETCH_GROUP_SETTINGS,
    graphql: {
      query: `query ($slug: String) {
        group (slug: $slug) {
          id
          accessibility
          avatarUrl
          bannerUrl
          description
          location
          invitePath
          name
          settings {
            allowGroupInvites
            publicMemberDirectory
          }
          slug
          visibility
          childGroups (first: 100) {
            items {
              id
              name
              avatarUrl
            }
          }
          moderators (first: 100) {
            hasMore
            items {
              id
              name
              avatarUrl
            }
          }
          parentGroups (first: 100) {
            items {
              id
              name
              avatarUrl
            }
          }
          pendingInvitations {
            hasMore
            items {
              id
              email
              createdAt
              lastSentAt
            }
          }
          questions {
            items {
              id
              text
            }
          }
        }
      }`,
      variables: {
        slug
      }
    },
    meta: {
      extractModel: 'Group'
    }
  }
}

export function updateGroupSettings (id, changes) {
  return {
    type: UPDATE_GROUP_SETTINGS,
    graphql: {
      query: `mutation ($id: ID, $changes: GroupInput) {
        updateGroupSettings(id: $id, changes: $changes) {
          id
        }
      }`,
      variables: {
        id, changes
      }
    },
    meta: {
      id,
      changes,
      optimistic: true
    }
  }
}

export function regenerateAccessCode (groupId) {
  return {
    type: REGENERATE_ACCESS_CODE,
    graphql: {
      query: `mutation ($groupId: ID) {
        regenerateAccessCode(groupId: $groupId) {
          id
          invitePath
        }
      }`,
      variables: {
        groupId
      }
    },
    meta: {
      extractModel: 'Group'
    }
  }
}

export function deleteGroup (id) {
  return {
    type: DELETE_GROUP,
    graphql: {
      query: `mutation ($id: ID) {
        deleteGroup(id: $id) {
          success
        }
      }`,
      variables: {
        id
      }
    },
    meta: {
      optimistic: true,
      id
    }
  }
}
