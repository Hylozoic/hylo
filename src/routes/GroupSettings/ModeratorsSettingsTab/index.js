import ModeratorsSettingsTab from './ModeratorsSettingsTab'
import connector from './ModeratorsSettingsTab.connector'
export default connector(ModeratorsSettingsTab)
