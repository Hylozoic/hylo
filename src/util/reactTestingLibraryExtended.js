import React from 'react'
import { MemoryRouter } from 'react-router'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import rootReducer from 'store/reducers'
import createMiddleware from 'store/middleware'

export function generateStore (history, initialState) {
  return createStore(rootReducer, initialState, createMiddleware(history))
}

export const AllTheProviders = (store) => ({ children }) => {
  return (
    <Provider store={store || createStore(() => {})}>
      <MemoryRouter>
        {children}
      </MemoryRouter>
    </Provider>
  )
}

// re-export everything

const rtl = require('react-testing-library')
const customRender = (ui, options, providersFunc) =>
  rtl.render(ui, { wrapper: providersFunc, ...options })

module.exports = {
  ...rtl,
  render: customRender
}
