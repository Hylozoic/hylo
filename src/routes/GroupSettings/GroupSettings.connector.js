import { push } from 'connected-react-router'
import { get } from 'lodash/fp'
import { connect } from 'react-redux'
import getGroupForCurrentRoute from 'store/selectors/getGroupForCurrentRoute'
import getRouteParam from 'store/selectors/getRouteParam'
import getCanModerate from 'store/selectors/getCanModerate'
import getMe from 'store/selectors/getMe'
import {
  fetchGroupSettings, updateGroupSettings, deleteGroup
} from './GroupSettings.store'
import { groupDeleteConfirmationUrl } from 'util/navigation'

export function mapStateToProps (state, props) {
  const slug = getRouteParam('groupSlug', state, props, false)
  const group = getGroupForCurrentRoute(state, props)
  const canModerate = getCanModerate(state, { group })

  return {
    canModerate,
    group,
    currentUser: getMe(state),
    slug
  }
}

export function mapDispatchToProps (dispatch, props) {
  return {
    deleteGroup: id => dispatch(deleteGroup(id)),
    fetchGroupSettingsMaker: slug => () => dispatch(fetchGroupSettings(slug)),
    goToGroupDeleteConfirmation: () => dispatch(push(groupDeleteConfirmationUrl())),
    updateGroupSettingsMaker: id => changes => dispatch(updateGroupSettings(id, changes))
  }
}

export function mergeProps (stateProps, dispatchProps, ownProps) {
  const { group, slug } = stateProps
  const {
    fetchGroupSettingsMaker, goToGroupDeleteConfirmation, updateGroupSettingsMaker
  } = dispatchProps
  let deleteGroup, fetchGroupSettings, updateGroupSettings

  if (slug) {
    fetchGroupSettings = fetchGroupSettingsMaker(slug)
  } else {
    fetchGroupSettings = () => {}
  }

  if (get('id', group)) {
    deleteGroup = () =>
      dispatchProps.deleteGroup(group.id)
        .then(({ error }) => {
          if (!error) {
            return goToGroupDeleteConfirmation()
          }
        })
    updateGroupSettings = updateGroupSettingsMaker(group.id)
  } else {
    deleteGroup = () => {}
    updateGroupSettings = () => {}
  }

  return {
    ...stateProps,
    ...dispatchProps,
    ...ownProps,
    deleteGroup,
    fetchGroupSettings,
    updateGroupSettings
  }
}

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)
