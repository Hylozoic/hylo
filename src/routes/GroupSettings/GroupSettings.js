import { get, compact } from 'lodash/fp'
import React, { useEffect } from 'react'
import { useTranslation } from 'react-i18next'
import AgreementsTab from './AgreementsTab'
import CustomViewsTab from './CustomViewsTab'
import DeleteSettingsTab from './DeleteSettingsTab'
import GroupSettingsTab from './GroupSettingsTab'
import ImportExportSettingsTab from './ImportExportSettingsTab'
import InviteSettingsTab from './InviteSettingsTab'
import MembershipRequestsTab from './MembershipRequestsTab'
import ModeratorsSettingsTab from './ModeratorsSettingsTab'
import PrivacySettingsTab from './PrivacySettingsTab'
import RelatedGroupsTab from './RelatedGroupsTab'
import ResponsibilitiesTab from './ResponsibilitiesTab'
import TopicsSettingsTab from './TopicsSettingsTab'
import ExportDataTab from './ExportDataTab'
import Loading from 'components/Loading'
import FullPageModal from 'routes/FullPageModal'
import { push } from 'connected-react-router'
import { Redirect } from 'react-router'
import { groupUrl } from 'util/navigation'
import { RESP_ADD_MEMBERS, RESP_ADMINISTRATION } from 'store/constants'

import './GroupSettings.scss'
import getResponsibilitiesForGroup from 'store/selectors/getResponsibilitiesForGroup'

// NOTE: This area is also rendered and shared with the mobile app.
// When making changes here or in any of the settings tabs please
// confirm accurate rendering and function in the related mobile area.
export default function GroupSettings ({
  addPostToCollection,
  canModerate,
  createCollection,
  currentUser,
  deleteGroup,
  fetchCollectionPosts,
  fetchCollectionPostsPending,
  fetchGroupSettings,
  fetchLocation,
  fetchPending,
  group,
  parentGroups,
  prerequisiteGroups,
  removePostFromCollection,
  reorderPostInCollection,
  updateGroupSettings,
  upload
}) {
  const slug = get('slug', group)
  const { t } = useTranslation()

  useEffect(() => {
    group && fetchGroupSettings()
  }, [slug])

  const responsibilities = getResponsibilitiesForGroup({ currentUser, groupId: group?.id }).map(r => r.title)
  if (!group) return <Loading />
  if (!canModerate && responsibilities.length === 0) return <Redirect to={groupUrl(slug)} />

  useEffect(() => {
    if (!canModerate && !responsibilities.includes(RESP_ADMINISTRATION) && responsibilities.includes(RESP_ADD_MEMBERS)) push(groupUrl(slug, 'settings/invite'))
  }, [])

  const overallSettings = {
    name: t('Settings'),
    path: groupUrl(slug, 'settings'),
    component: <GroupSettingsTab
      fetchLocation={fetchLocation}
      fetchPending={fetchPending}
      currentUser={currentUser}
      group={group}
      parentGroups={parentGroups}
      prerequisiteGroups={prerequisiteGroups}
      updateGroupSettings={updateGroupSettings}
    />
  }

  const agreementSettings = {
    name: t('Agreements'),
    path: groupUrl(slug, 'settings/agreements'),
    component: <AgreementsTab group={group} />
  }

  const responsibilitiesSettings = {
    name: t('Responsibilities'),
    path: groupUrl(slug, 'settings/responsibilities'),
    component: <ResponsibilitiesTab groupId={group.id} group={group} slug={group.slug} />
  }

  const rolesSettings = {
    name: t('Roles & Badges'),
    path: groupUrl(slug, 'settings/roles'),
    component: <ModeratorsSettingsTab groupId={group.id} group={group} slug={group.slug} />
  }

  const accessSettings = {
    name: t('Privacy & Access'),
    path: groupUrl(slug, 'settings/privacy'),
    component: <PrivacySettingsTab group={group} slug={group.slug} updateGroupSettings={updateGroupSettings} parentGroups={parentGroups} fetchPending={fetchPending} />
  }

  const customViewsSettings = {
    name: t('Custom Views'),
    path: groupUrl(slug, 'settings/views'),
    component: <CustomViewsTab
      group={group}
      addPostToCollection={addPostToCollection}
      createCollection={createCollection}
      fetchCollectionPosts={fetchCollectionPosts}
      fetchCollectionPostsPending={fetchCollectionPostsPending}
      fetchPending={fetchPending}
      removePostFromCollection={removePostFromCollection}
      reorderPostInCollection={reorderPostInCollection}
      updateGroupSettings={updateGroupSettings}
    />
  }

  const topicsSettings = {
    name: t('Topics'),
    path: groupUrl(slug, 'settings/topics'),
    component: <TopicsSettingsTab group={group} />
  }

  const inviteSettings = {
    name: t('Invite'),
    path: groupUrl(slug, 'settings/invite'),
    component: <InviteSettingsTab group={group} />
  }

  const joinRequestSettings = {
    name: t('Join Requests'),
    path: groupUrl(slug, 'settings/requests'),
    component: <MembershipRequestsTab
      group={group}
      currentUser={currentUser}
    />
  }

  const relatedGroupsSettings = {
    name: t('Related Groups'),
    path: groupUrl(slug, 'settings/relationships'),
    component: <RelatedGroupsTab
      group={group}
      currentUser={currentUser}
    />
  }

  const importSettings = {
    name: '',
    path: groupUrl(slug, 'settings/import'),
    component: <ImportExportSettingsTab
      group={group}
      upload={upload}
    />
  }

  const exportSettings = {
    name: t('Export Data'),
    path: groupUrl(slug, 'settings/export'),
    component: <ExportDataTab
      group={group}
    />
  }

  const deleteSettings = {
    name: t('Delete'),
    path: groupUrl(slug, 'settings/delete'),
    component: <DeleteSettingsTab group={group} deleteGroup={deleteGroup} />
  }

  return (
    <FullPageModal
      goToOnClose={groupUrl(slug)}
      content={compact([
        canModerate || responsibilities.includes(RESP_ADMINISTRATION) ? overallSettings : null,
        canModerate || responsibilities.includes(RESP_ADMINISTRATION) ? agreementSettings : null,
        canModerate || responsibilities.includes(RESP_ADMINISTRATION) ? responsibilitiesSettings : null,
        canModerate || responsibilities.includes(RESP_ADMINISTRATION) ? rolesSettings : null,
        canModerate || responsibilities.includes(RESP_ADMINISTRATION) ? accessSettings : null,
        canModerate || responsibilities.includes(RESP_ADMINISTRATION) ? customViewsSettings : null,
        canModerate || responsibilities.includes(RESP_ADMINISTRATION) ? topicsSettings : null,
        canModerate || responsibilities.includes(RESP_ADD_MEMBERS) ? inviteSettings : null,
        canModerate || responsibilities.includes(RESP_ADD_MEMBERS) ? joinRequestSettings : null,
        canModerate || responsibilities.includes(RESP_ADMINISTRATION) ? relatedGroupsSettings : null,
        canModerate || responsibilities.includes(RESP_ADMINISTRATION) ? importSettings : null,
        canModerate || responsibilities.includes(RESP_ADMINISTRATION) ? exportSettings : null,
        canModerate ? deleteSettings : null
      ])}
    />
  )
}
