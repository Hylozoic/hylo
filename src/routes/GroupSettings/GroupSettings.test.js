import GroupSettings from './GroupSettings'
import React from 'react'
import { AllTheProviders, render } from 'util/testing/reactTestingLibraryExtended'

jest.mock('react-i18next', () => ({
  ...jest.requireActual('react-i18next'),
  useTranslation: (domain) => {
    return {
      t: (str) => str,
      i18n: {
        changeLanguage: () => new Promise(() => {})
      }
    }
  }
}))

it('renders correctly with no group', () => {
  const { asFragment } = render(<GroupSettings fetchGroupSettings={jest.fn()} />, { wrapper: AllTheProviders() })
  expect(asFragment()).toMatchSnapshot()
})

it('renders a redirect if you can not moderate', () => {
  const group = { id: 1, slug: 'foo', name: 'Foomunity' }
  const { asFragment } = render(<GroupSettings fetchGroupSettings={jest.fn()} group={group} />, { wrapper: AllTheProviders() })
  expect(asFragment()).toMatchSnapshot()
})

it('renders correctly with a group', () => {
  const group = { id: 1, slug: 'foo', name: 'Foomunity' }
  const { asFragment } = render(<GroupSettings fetchGroupSettings={jest.fn()} group={group} />, { wrapper: AllTheProviders() })
  expect(asFragment()).toMatchSnapshot()
})
