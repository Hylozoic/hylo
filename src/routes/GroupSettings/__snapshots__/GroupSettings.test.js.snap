// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a redirect if you can not moderate 1`] = `
<Redirect
  to="/c/foo"
/>
`;

exports[`renders correctly with a group 1`] = `
<withRouter(Connect(FullPageModal))
  content={
    Array [
      Object {
        "component": <GroupSettingsTab
          group={
            Object {
              "id": 1,
              "name": "Foomunity",
              "slug": "foo",
            }
          }
        />,
        "name": "Settings",
        "path": "/c/foo/settings",
      },
      Object {
<<<<<<< HEAD:src/routes/GroupSettings/__snapshots__/GroupSettings.test.js.snap
        "component": <Connect(ModeratorsSettingsTab)
          groupId={1}
=======
        "component": <Memo(Connect(ModeratorsSettingsTab))
          communityId={1}
>>>>>>> dev:src/routes/CommunitySettings/__snapshots__/CommunitySettings.test.js.snap
          slug="foo"
        />,
        "name": "Moderators",
        "path": "/c/foo/settings/moderators",
      },
      Object {
<<<<<<< HEAD:src/routes/GroupSettings/__snapshots__/GroupSettings.test.js.snap
        "component": <Connect(TopicsSettingsTab)
          group={
=======
        "component": <Memo(Connect(TopicsSettingsTab))
          community={
>>>>>>> dev:src/routes/CommunitySettings/__snapshots__/CommunitySettings.test.js.snap
            Object {
              "id": 1,
              "name": "Foomunity",
              "slug": "foo",
            }
          }
        />,
        "name": "Topics",
        "path": "/c/foo/settings/topics",
      },
      Object {
<<<<<<< HEAD:src/routes/GroupSettings/__snapshots__/GroupSettings.test.js.snap
        "component": <Connect(InviteSettingsTab)
          group={
=======
        "component": <Memo(Connect(InviteSettingsTab))
          community={
>>>>>>> dev:src/routes/CommunitySettings/__snapshots__/CommunitySettings.test.js.snap
            Object {
              "id": 1,
              "name": "Foomunity",
              "slug": "foo",
            }
          }
        />,
        "name": "Invite",
        "path": "/c/foo/settings/invite",
      },
      Object {
<<<<<<< HEAD:src/routes/GroupSettings/__snapshots__/GroupSettings.test.js.snap
        "component": <Connect(MembershipRequestsTab)
          group={
=======
        "component": <Memo(Connect(MembershipRequestsTab))
          community={
>>>>>>> dev:src/routes/CommunitySettings/__snapshots__/CommunitySettings.test.js.snap
            Object {
              "id": 1,
              "name": "Foomunity",
              "slug": "foo",
            }
          }
        />,
        "name": "Join Requests",
        "path": "/c/foo/settings/requests",
      },
      Object {
        "component": <ImportExportSettingsTab
          group={
            Object {
              "id": 1,
              "name": "Foomunity",
              "slug": "foo",
            }
          }
        />,
        "name": "",
        "path": "/c/foo/settings/import",
      },
      Object {
        "component": <DeleteSettingsTab
          group={
            Object {
              "id": 1,
              "name": "Foomunity",
              "slug": "foo",
            }
          }
        />,
        "name": "Delete",
        "path": "/c/foo/settings/delete",
      },
    ]
  }
  goToOnClose="/c/foo"
/>
`;

exports[`renders correctly with no group 1`] = `<Loading />`;
